### PR TITLE
CI: Meson CI(s) on GitHub.

### PR DESCRIPTION
**Pull request by mobin-2008**
Previously Meson based CI hosted on SourceHut. That was great but sr.ht doesn't have good support for integrating with GitHub due to sunset of 'dispatch.sr.ht`. So I moving my workflows from there to GitHub for better Integration.

These are designed to be in main repo (davmac314/dinit) but these aren't running by default. These will be ran when a PR in main repo has "meson" label.

Regards

`Signed-off-by: Mobin "Hojjat" Aydinfar <mobin@mobintestserver.ir>`

### CODE REVIEW

**Comment by mobin-2008** [comment]
As see, Those just skipped.

**Comment by mobin-2008** [comment]
Hmm, seems like It's need a another commit after assign "meson" label :/

**Comment by davmac314** [comment]
Since `workflow-dispatch` is a trigger, it can be run manually if needed (i.e. for new pull requests where the label isn't set yet).

**Comment by mobin-2008** [comment]
> Since `workflow-dispatch` is a trigger, it can be run manually if needed (i.e. for new pull requests where the label isn't set yet).

`workflow-dispatch` is for local branches, It doesn't work for PR(s).

**Comment by davmac314** [comment]
Ah, you're right. That's an annoying limitation.

**Comment by mobin-2008** [comment]
I changed code to run meson workflows when a "rerun" requested. I think it's ready now.

File: .github/workflows/meson_ci.yml [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+      - '!doc/manpages/README'
+      - '!doc/manpages/generate-html.sh'
+      - '!doc/linux/**'
+  workflow_dispatch:
+
+jobs:
+
+  Debian-bullseye_build:
+
+    if: ${{ github.run_attempt >= 2 || contains(github.event.pull_request.labels.*.name, 'meson') }}
```
Comment (davmac314) [inline .github/workflows/meson_ci.yml:35]: So *any* rerun of CI will always run the meson CI builds? Is that what you intended?

Comment (davmac314) [reply .github/workflows/meson_ci.yml:35]: Could we just leave the condition as it was ("pull request label contains 'meson'") and then if we want to run meson CI for a given pull request just add the label and then re-run CI?

Comment (mobin-2008) [reply .github/workflows/meson_ci.yml:35]: > So _any_ rerun of CI will always run the meson CI builds? Is that what you intended?

Any rerun of **Meson** CI will always run the meson CI builds.

Comment (mobin-2008) [reply .github/workflows/meson_ci.yml:35]: > Could we just leave the condition as it was ("pull request label contains 'meson'") and then if we want to run meson CI for a given pull request just add the label and then re-run CI?

I tested it before but it's need an another commit to work.

Comment (davmac314) [reply .github/workflows/meson_ci.yml:35]: > Any rerun of Meson CI will always run the meson CI builds.

But the condition is "or":

```
    if: ${{ github.run_attempt >= 2 || contains(github.event.pull_request.labels.*.name, 'meson') }}
```

If this is the 2nd run, OR the pull event contains 'meson'

Comment (mobin-2008) [reply .github/workflows/meson_ci.yml:35]: **Regular CI** and **Meson based CI** are two different workflows and rerun of one of them, doesn't rerun other workflow.

Comment (davmac314) [reply .github/workflows/meson_ci.yml:35]: Alright, so how can you re-run the meson workflow if it never ran the first time?

Comment (mobin-2008) [reply .github/workflows/meson_ci.yml:35]: 1. a PR is related to meson but it doesn't have meson label.
2. Meson CI skipped for first time.
3. one of the maintainers rerun skipped workflow. it's will make `github.run_attempt` = 2.
4. Meson CI will run for that PR. Also we add meson label for auto meson ci.

Comment (davmac314) [reply .github/workflows/meson_ci.yml:35]: Why not:

1. a PR is related to meson but it doesn't have meson label.
2. Meson CI skipped for first time.
3. one of the maintainers adds the  meson label
4. maintainer then reruns the skipped workflow

Doesn't that achieve the same thing?

Comment (mobin-2008) [reply .github/workflows/meson_ci.yml:35]: 5. Unfortunately it's skipped again due to some reasons.

Comment (davmac314) [reply .github/workflows/meson_ci.yml:35]: > > So _any_ rerun of CI will always run the meson CI builds? Is that what you intended?
> 
> Any rerun of **Meson** CI will always run the meson CI builds.

I don't see how to run just Meson CI. Github UI says "rerun all jobs". It's not specific to a workflow.

image

Comment (davmac314) [reply .github/workflows/meson_ci.yml:35]: So if the condition is as it is currently, re-running jobs will run the meson CI workflow even if it isn't a meson-related pull request.

Comment (davmac314) [reply .github/workflows/meson_ci.yml:35]: Hmm, maybe I'm wrong. Maybe it separates different workflows into different jobs. But I want to check more, will need to do that later.

Comment (mobin-2008) [reply .github/workflows/meson_ci.yml:35]: For sure, I reran Regular CI for your last commit. As you see It doesn't trigger Meson based CI.
https://github.com/davmac314/dinit/actions/runs/5115941484

Comment (davmac314) [reply .github/workflows/meson_ci.yml:35]: Ok, sounds good.

Can you please put a comment in the file explaining all this (just a couple of sentences is fine).

Can we make it run on any change to master (once merged/pushed)? I think it's good to know if meson build does break

Comment (davmac314) [reply .github/workflows/meson_ci.yml:35]: Details: https://blog.mergify.com/running-github-actions-only-on-certain-pull-requests/

Comment (mobin-2008) [reply .github/workflows/meson_ci.yml:35]: I think we need to bypass that check in workflow for pushes, No? Currently it needs a re-run request to run.

Comment (davmac314) [reply .github/workflows/meson_ci.yml]: > Currently it needs a re-run request to run.

I think with the correct dispatch (`on: pull_request: types: - labeled') it won't need a re-run request.

But yes, the conditions may need to be different for a pull request vs push to master.

Comment (mobin-2008) [reply .github/workflows/meson_ci.yml]: But seems like it doesn't work on newer commits!

Comment (davmac314) [reply .github/workflows/meson_ci.yml]: you need it dispatched on the default event types too - `opened`, `reopened` and `synchronize`. If the event type is `label` it only triggers at the time the label is added.

**Pull request merged**